### PR TITLE
add option for SameSite cookies

### PR DESCRIPTION
**Pull request by r0fls**
This MR adds support for the `SameSite=Strict` cookie directive as described here:
https://tools.ietf.org/html/draft-west-first-party-cookies-07

### CODE REVIEW

**Comment by iangcarroll** [comment]
Whoops, I duplicated this in #1059. Any reason you don't support `Lax`? `Strict` is very restrictive and impractical for many things where you need sessions on every page load.

**Comment by r0fls** [comment]
@iangcarroll no, there is no reason not to add that. Good call 👍  I can close this but you should an a unit test to your implementation.

File: werkzeug/http.py [hunk 1]
```diff
@@ -937,7 +937,7 @@ def _parse_pairs():
 
 def dump_cookie(key, value='', max_age=None, expires=None, path='/',
                 domain=None, secure=False, httponly=False,
-                charset='utf-8', sync_expires=True):
+                charset='utf-8', sync_expires=True, samesite=None):
```
Comment (ThiefMaster) [inline werkzeug/http.py]: docstring needs to be updated

File: werkzeug/http.py [hunk 2]
```diff
@@ -998,6 +1001,7 @@ def dump_cookie(key, value='', max_age=None, expires=None, path='/',
                     (b'Max-Age', max_age, False),
                     (b'Secure', secure, None),
                     (b'HttpOnly', httponly, None),
+                    (b'SameSite', samesite, False),
```
Comment (ThiefMaster) [inline werkzeug/http.py]: q should be `None` according to the docstring and code below.

File: werkzeug/http.py [hunk 3]
```diff
@@ -988,6 +988,9 @@ def dump_cookie(key, value='', max_age=None, expires=None, path='/',
     elif max_age is not None and sync_expires:
         expires = to_bytes(cookie_date(time() + max_age))
 
+    if samesite:
+        samesite = 'Strict'
```
Comment (ThiefMaster) [inline werkzeug/http.py]: This is broken. You have an option that appears to take a boolean or a string but if it's truthy you always set it to strict. I'd go for this:

```
if samesite not in (None, 'Strict', 'Lax'):
    raise ValueError("invalid samesite value; must be 'Strict', 'Lax' or None")
```

or this:

```
if samesite and samesite not in ('Strict', 'Lax'):
    raise ValueError("invalid samesite value; must be 'Strict', 'Lax' or None")
```

File: werkzeug/wrappers.py [hunk 4]
```diff
@@ -1003,7 +1003,8 @@ def iter_encoded(self):
         return _iter_encoded(self.response, self.charset)
 
     def set_cookie(self, key, value='', max_age=None, expires=None,
-                   path='/', domain=None, secure=False, httponly=False):
+                   path='/', domain=None, secure=False, httponly=False,
+                   samesite=None):
```
Comment (ThiefMaster) [inline werkzeug/wrappers.py]: docstring needs to be updated

File: tests/test_http.py [hunk 5]
```diff
@@ -384,6 +384,10 @@ def test_cookie_domain_resolving(self):
         val = http.dump_cookie('foo', 'bar', domain=u'\N{SNOWMAN}.com')
         strict_eq(val, 'foo=bar; Domain=xn--n3h.com; Path=/')
 
+    def test_cookie_samesite(self):
+        val = http.dump_cookie('foo', 'bar', samesite=True)
```
Comment (ThiefMaster) [inline tests/test_http.py]: see other comments; `True` does not make much sense for this option so this test should cover both none, strict and lax mode.

Comment (ThiefMaster) [inline tests/test_http.py]: You can use pytest parametrization to test all 3 cases.

**Comment by acheronflow** [comment]
+1 on this, I was just looking how to do that for my Flask apps. May I suggest the check for given "samestrict" value (through set_cookie) be case insensitive?

**Comment by ThiefMaster** [comment]
Good idea, updating my comment with the code suggestion

**Comment by mitsuhiko** [comment]
This is a duplicate.

**Comment by mitsuhiko** [comment]
This is a duplicate.

**Pull request closed without merging**